src/stac_fastapi/collection_discovery/core.py

```diff
@@ -0,0 +1,321 @@
+    Item,
+    ItemCollection,
+    LandingPage,
+)
+
+BASE_CONFORMANCE_CLASSES = [
+    STACConformanceClasses.CORE,
+    STACConformanceClasses.COLLECTIONS,
+    OAFConformanceClasses.CORE,
+    OAFConformanceClasses.OPEN_API,
+]
+
+logger = logging.getLogger(__name__)
+
+
+class CollectionSearchClient(AsyncBaseCoreClient):
+    base_conformance_classes: List[str] = attr.ib(
+        factory=lambda: BASE_CONFORMANCE_CLASSES
+    )
+
+    def _encode_token(self, state: Dict) -> str:
+        """Encode pagination state into a token."""
+        json_str = json.dumps(state, separators=(",", ":"))
+        return base64.urlsafe_b64encode(json_str.encode()).decode()
+
+    def _decode_token(self, token: str) -> Dict:
+        """Decode token back into pagination state."""
+        json_str = base64.urlsafe_b64decode(token.encode()).decode()
+        return json.loads(json_str)
+
+    async def all_collections(
+        self,
+        request: Request,
+        token: Optional[str] = None,
+        bbox: Optional[BBox] = None,
+        datetime: Optional[str] = None,
+        limit: Optional[int] = None,
+        fields: Optional[List[str]] = None,
+        sortby: Optional[str] = None,
+        filter_expr: Optional[str] = None,
+        filter_lang: Optional[str] = None,
+        q: Optional[List[str]] = None,
+        **kwargs,
+    ) -> Collections:
+        _bbox = ",".join(str(coord) for coord in bbox) if bbox else None
+
+        params = {
+            key: value
+            for key, value in {
+                "bbox": _bbox,
+                "datetime": datetime,
+                "limit": limit,
+                "fields": fields,
+                "sortby": sortby,
+                "filter_expr": filter_expr,
+                "filter_lang": filter_lang,
+                "q": q,
+            }.items()
+            if value is not None
+        }
+        collections = []
+        links = [
+            {
+                "rel": "self",
+                "href": str(request.url),
+            }
+        ]
+        param_str = unquote(urlencode(params, True))
+
+        if token:
+            search_state = self._decode_token(token)
+            logger.info("Continuing collection search with token pagination")
+        else:
+            search_state = {
+                "current": {
+                    api: f"{api}/collections?{param_str}"
+                    for api in request.app.state.settings.child_api_urls
+                },
+                "is_first_page": True,
+            }
+            logger.info(
+                "Starting new collection search across "
+                f"{len(request.app.state.settings.child_api_urls)} APIs"
+            )
+
+        new_state: Dict[str, Any] = {
+            "current": {},
+            "previous": {},
+            "next": {},
+            "is_first_page": False,
+        }
+
+        async def fetch_api_data(
+            client, api: str, url: str
+        ) -> Tuple[str, Dict[str, Any]]:
```

Comment:
Nit: we can just use `tuple` and `dict` as of Python 3.10.

```suggestion
        ) -> tuple[str, dict[str, Any]]:
```